librz/analysis/var_global.c

```diff
@@ -0,0 +1,432 @@
+// SPDX-FileCopyrightText: 2021 Basstorm <basstorm@nyist.edu.cn>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+#include <rz_list.h>
+
+/**
+ * \brief Create a new instance of global variable
+ * 
+ * \param name variable name
+ * \param addr variable address
+ * \param comment variable comment
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_OWN RzAnalysisVarGlobal *rz_analysis_var_global_new(RZ_NONNULL const char *name, ut64 addr, RZ_NULLABLE const char *comment) {
+	rz_return_val_if_fail(name && addr, NULL);
+	RzAnalysisVarGlobal *glob = RZ_NEW0(RzAnalysisVarGlobal);
+	if (!glob) {
+		return NULL;
+	}
+	glob->name = strdup(name);
+	glob->addr = addr;
+	if (comment) {
+		glob->comment = strdup(comment);
+	}
+	return glob;
+}
+
+/**
+ * \brief Add the global variable into hashtable
+ * 
+ * \param analysis RzAnalysis
+ * \param global_var Global variable instance
+ * \return true if succeed
+ */
+RZ_API RZ_OWN bool rz_analysis_var_global_add(RzAnalysis *analysis, RzAnalysisVarGlobal *global_var) {
+	rz_return_val_if_fail(analysis && global_var, false);
+	if (rz_analysis_var_global_get_byaddr(analysis, global_var->addr)) {
+		RZ_LOG_ERROR("Global variable at 0x%" PFMT64x " is already exist!\n", global_var->addr);
+		return false;
+	}
+	return ht_pp_insert(analysis->ht_global_var, global_var->name, global_var);
+}
+
+/**
+ * \brief Free the global variable instance
+ * 
+ * \param glob Global variable instance
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_free(RzAnalysisVarGlobal *glob) {
+	if (!glob) {
+		return;
+	}
+
+	RZ_FREE(glob->name);
+	RZ_FREE(glob->comment);
+	rz_type_free(glob->type);
+	rz_analysis_var_global_clear_accesses(glob);
+	rz_vector_fini(&glob->constraints);
+	RZ_FREE(glob);
+}
+
+/**
+ * \brief Delete and free the global variable by its name
+ * 
+ * \param analysis RzAnalysis
+ * \param name Global Variable name
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_delete_byname(RzAnalysis *analysis, const char *name) {
+	rz_return_val_if_fail(analysis && name, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byname(analysis, name);
+	if (!glob) {
+		RZ_LOG_ERROR("No such global variable!\n");
```

Comment:
I don't think this is necessary... I think we should not print anything in such cases and it should be the caller of the API to choose what to write, if write at all something.